sw/device/lib/base/memory_perftest.c

```diff
@@ -0,0 +1,308 @@
+  }
+  CHECK(test->func != NULL);
+
+  uint64_t start_cycles = read_cycle_counter();
+  test->func(context);
+  uint64_t end_cycles = read_cycle_counter();
+
+  // Check that executing the test function will take more than zero cycles.
+  // Practically, we should never see the 64-bit cycle counter overflow. Even if
+  // it took 12 hours to run all the perf tests, the clock would have to run at
+  // 427 THz to overflow the counter (2**64 cycles / (12 * 60 * 60) seconds).
+  CHECK(start_cycles < end_cycles);
+  return end_cycles - start_cycles;
+}
+
+static inline uint64_t read_cycle_counter() {
+#ifdef OT_PLATFORM_RV32
+  // We cannot read the entire 64-bit mcycle counter at once on RISCV32.
+  uint32_t mcycleh, mcycleh2, mcycle_lower;
+  do {
+    CSR_READ(CSR_REG_MCYCLEH, &mcycleh);
+    CSR_READ(CSR_REG_MCYCLE, &mcycle_lower);
+    CSR_READ(CSR_REG_MCYCLEH, &mcycleh2);
+  } while (mcycleh != mcycleh2);
+
+  return ((uint64_t)mcycleh << 32) | mcycle_lower;
+#else
+  static_assert(false, "Not implemented");
+#endif
+}
+
+static inline bool compute_statistics(const uint64_t *durations,
+                                      size_t num_durations, uint32_t *out_mean,
+                                      uint32_t *out_variance) {
+  uint64_t duration_sum = 0;
+  for (size_t rep = 0; rep < num_durations; ++rep) {
+    if (duration_sum > UINT64_MAX - durations[rep]) {
+      LOG_WARNING("Summing durations would overflow.");
+      return false;
+    }
+    duration_sum += durations[rep];
+  }
+  CHECK(num_durations > 0);
+  const uint64_t duration_mean =
+      udiv64_slow(duration_sum, num_durations, /*rem_out=*/NULL);
+
+  uint64_t sum_squared_differences = 0;
+  for (size_t rep = 0; rep < num_durations; ++rep) {
+    uint64_t difference;
+    if (durations[rep] > duration_mean) {
+      difference = durations[rep] - duration_mean;
+    } else if (durations[rep] < duration_mean) {
+      difference = duration_mean - durations[rep];
+    } else {
+      continue;
+    }
+    if (difference > udiv64_slow(UINT64_MAX, difference, /*rem_out=*/NULL)) {
+      LOG_WARNING("Squaring difference would overflow.");
+      return false;
+    }
+    if (sum_squared_differences > UINT64_MAX - difference * difference) {
+      LOG_WARNING("Adding squared difference to sum would overflow.");
+      return false;
+    }
+    sum_squared_differences += difference * difference;
+  }
+
+  CHECK(num_durations > 1);
+  const uint64_t duration_variance = udiv64_slow(
+      sum_squared_differences, (uint64_t)num_durations - 1, /*rem_out=*/NULL);
+
+  if (duration_mean > UINT32_MAX) {
+    LOG_WARNING("Computed mean cannot fit in 32 bits");
+    return false;
+  }
+  if (duration_variance > UINT32_MAX) {
+    LOG_WARNING("Computed variance cannot fit in 32 bits");
+    return false;
+  }
+  *out_mean = (uint32_t)duration_mean;
+  *out_variance = (uint32_t)duration_variance;
+  return true;
+}
+
+static inline void fill_buf_arbitrary_values(uint8_t *buf, size_t len) {
```

Comment:
Does `inline` do anything here?  You're taking the address of this function when you set up the table of benchmarks later, so this function cannot be inline for those cases.

It seems like `inline` has been added to everything.  I _think_ the C compiler is going to make its own decisions about whether to inline functions based on the optimization level; I'm not certain the `inline` keyword is useful for these functions.

Reply:
> Does inline do anything here? You're taking the address of this function when you set up the table of benchmarks later, so this function cannot be inline for those cases.

Nope, but it made sense before I refactored this and started taking the address of the function. Good catch.

> It seems like inline has been added to everything. I think the C compiler is going to make its own decisions about whether to inline functions based on the optimization level; I'm not certain the inline keyword is useful for these functions.

Maybe I was a bit overzealous :P. I was playing with `objdump` and noticed that `inline` was having an effect, so I tried to specify it when applicable.

Playing around in Compiler Explorer [^0] confirms your suspicion about optimization and inlining. I saw that when `-O2` is specified, both `inline` and not-necessarily-`inline` static functions are inlined. Without any `-O` flag, they are not inlined.

On that note, I guess we should be specifying an optimization level in the new functest's cflags. Not much point in perftesting code that is not optimized the same way it would be in production.

[^0]:  https://godbolt.org/z/dbxGvhMc5